### PR TITLE
[CPU] Optimize infer request outputs memory sharing

### PR DESCRIPTION
**Pull request by ceciliapeng2011**
### Details:
 - *change the default case resolution policy so that it resolves cyclic inplace to downstream instead of upstream for outputs.*
 - *this is a more optimal decision as more cases are going to reuse infer request's memory in PullOutputData.*

### Tickets:
 - *116823*
 - *120566*

This reopens legacy PR https://github.com/openvinotoolkit/openvino/pull/19684 as more models requires this feature.

### CODE REVIEW

File: src/plugins/intel_cpu/src/graph.cpp [hunk 1]
```diff
@@ -1810,14 +1810,66 @@ void Graph::resolveInPlaceDirection(const NodePtr& node) const {
...
                     node->initDescriptor(config);
                 } else if (InplaceDirectionType::DOWN == result) {
                     auto config = node->getSelectedPrimitiveDescriptor()->getConfig();
                     config.outConfs[inPlaceInpPort].inPlace(-1);
                     node->initDescriptor(config);
+                } else if (InplaceDirectionType::NONE == result) {
+                    // resolve cyclic inplace to downstream instead of upstream for the node
+                    // when there is only one output referencing to the edges of it,
+                    // thus benefits zero-copy of outputs.
+                    int numConflicts = 0;
```
Comment (maxnick) [inline src/plugins/intel_cpu/src/graph.cpp:1825]: It looks like `numConflicts` is never decremented. Better to use unsigned type, e.g. `size_t`.

File: src/plugins/intel_cpu/tests/functional/custom/subgraph_tests/src/inplace_resolve_io.cpp [hunk 2]
```diff
@@ -0,0 +1,333 @@
...
+        auto concat = std::make_shared<ov::op::v0::Concat>(ov::OutputVector{rnnseq->outputs()[1], soft_max}, 0);  // Ho [batch_size, num_directions, hidden_size]
+        auto result_0 = std::make_shared<ov::op::v0::Result>(concat);
+
+        auto reshape_param_1 = ov::op::v0::Constant::create(ov::element::i32, {1}, {1});
+        auto reshape_1 = std::make_shared<ov::op::v0::Squeeze>(rnnseq->outputs()[0], reshape_param_1);  // Y [batch_size, num_directions, seq_len, hidden_size]
+        auto result_1 = std::make_shared<ov::op::v0::Result>(reshape_1);
+
+        function = std::make_shared<ov::Model>(ov::ResultVector{result_0, result_1}, params, "Subgraph0");
+
+        ov::pass::Serialize serializer("Subgraph0.xml", "Subgraph0.bin");
```
Comment (maxnick) [inline src/plugins/intel_cpu/tests/functional/custom/subgraph_tests/src/inplace_resolve_io.cpp:102]: Please clean up debug code.

File: src/plugins/intel_cpu/tests/functional/custom/subgraph_tests/src/inplace_resolve_io.cpp [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+                        |
+                      Softmax
+                     /     \
+                    /       \
+                   Add     Reshape0
+                    |         |
+                    |         |
+                 Result0   Result1
+
+expect edge Reshape1->Result1 to be referenced by its upstreams, instead of referencing to its upstreams.
```
Comment (maxnick) [inline src/plugins/intel_cpu/tests/functional/custom/subgraph_tests/src/inplace_resolve_io.cpp:175]: The comment is not very clear, could you please rephrase it?

File: src/plugins/intel_cpu/tests/functional/custom/subgraph_tests/src/inplace_resolve_io.cpp [hunk 4]
```diff
@@ -0,0 +1,333 @@
...
+        auto result_0 = std::make_shared<ov::op::v0::Result>(add_0);    // dummy output
+
+        auto reshape_param_0 = ov::op::v0::Constant::create(ov::element::i32, {1}, {0});
+        auto reshape_0 = std::make_shared<ov::op::v0::Unsqueeze>(soft_max, reshape_param_0);
+        auto result_1 = std::make_shared<ov::op::v0::Result>(reshape_0);    // target output
+
+        ov::ResultVector results = {result_0, result_1};
+        function = std::make_shared<ov::Model>(results, params, "Subgraph1");
+
+        ov::pass::Serialize serializer("Subgraph1.xml", "Subgraph1.bin");
```
Comment (maxnick) [inline src/plugins/intel_cpu/tests/functional/custom/subgraph_tests/src/inplace_resolve_io.cpp:201]: Please remove debug code.

File: src/plugins/intel_cpu/tests/functional/custom/subgraph_tests/src/inplace_resolve_io.cpp [hunk 5]
```diff
@@ -0,0 +1,333 @@
...
+        auto result_2 = std::make_shared<ov::op::v0::Result>(reshape_0);    // dummy output
+
+        auto reshape_param_1 = ov::op::v0::Constant::create(ov::element::i32, {1}, {0});
+        auto reshape_1 = std::make_shared<ov::op::v0::Unsqueeze>(reshape_0, reshape_param_1);
+        auto result_1 = std::make_shared<ov::op::v0::Result>(reshape_1);    // target output
+
+        ov::ResultVector results = {result_0, result_1, result_2};
+        function = std::make_shared<ov::Model>(results, params, "Subgraph2");
+
+        ov::pass::Serialize serializer("Subgraph2.xml", "Subgraph2.bin");
```
Comment (maxnick) [inline src/plugins/intel_cpu/tests/functional/custom/subgraph_tests/src/inplace_resolve_io.cpp:257]: Please remove debug code.

**Review by maxnick — CHANGES_REQUESTED** [review]

File: src/plugins/intel_cpu/src/graph.cpp (hunk 1 above)
Comment (ceciliapeng2011) [reply src/plugins/intel_cpu/src/graph.cpp]: done

File: src/plugins/intel_cpu/tests/functional/custom/subgraph_tests/src/inplace_resolve_io.cpp (hunk 2 above)
Comment (ceciliapeng2011) [reply src/plugins/intel_cpu/tests/functional/custom/subgraph_tests/src/inplace_resolve_io.cpp]: done

File: src/plugins/intel_cpu/tests/functional/custom/subgraph_tests/src/inplace_resolve_io.cpp (hunk 3 above)
Comment (ceciliapeng2011) [reply src/plugins/intel_cpu/tests/functional/custom/subgraph_tests/src/inplace_resolve_io.cpp]: done

File: src/plugins/intel_cpu/tests/functional/custom/subgraph_tests/src/inplace_resolve_io.cpp (hunk 5 above)
Comment (ceciliapeng2011) [reply src/plugins/intel_cpu/tests/functional/custom/subgraph_tests/src/inplace_resolve_io.cpp]: done

File: src/plugins/intel_cpu/tests/functional/custom/subgraph_tests/src/inplace_resolve_io.cpp
Comment (maxnick) [inline src/plugins/intel_cpu/tests/functional/custom/subgraph_tests/src/inplace_resolve_io.cpp:1]: While these tests are definitely good to have as smoke tasting. They only check the correctness of the obtained results, while they don't test the functionality introduced. We can remove all the changes introduced in this PR and these tests will still pass. Thus, on top of this subgraph tests, we need to have unit tests, which check that the memory is really shared in the expected way.

Comment (ceciliapeng2011) [reply src/plugins/intel_cpu/tests/functional/custom/subgraph_tests/src/inplace_resolve_io.cpp:1]: yes, will do it.

File: src/plugins/intel_cpu/tests/unit/graph/inplace_resolve_io.cpp [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+
+#include "common_test_utils/node_builders/constant.hpp"
+#include "common_test_utils/node_builders/eltwise.hpp"
+#include "common_test_utils/ov_tensor_utils.hpp"
+
+using namespace ov::intel_cpu;
+using namespace ov::op;
+
+// helper to check the inplace direction of a node with a part of its name
+void CheckInplaceDirection(const std::shared_ptr<Graph> graph, const std::string &patial_name, size_t inport, const Edge::LOOK undesiredDirection) {
```
Comment (maxnick) [inline src/plugins/intel_cpu/tests/unit/graph/inplace_resolve_io.cpp:24]: Isn't it better to check for "desired" direction?

File: src/plugins/intel_cpu/tests/functional/custom/subgraph_tests/src/inplace_resolve_io.cpp
Comment (ceciliapeng2011) [reply src/plugins/intel_cpu/tests/functional/custom/subgraph_tests/src/inplace_resolve_io.cpp:1]: done

File: src/plugins/intel_cpu/tests/unit/graph/inplace_resolve_io.cpp (hunk 6 above)
Comment (ceciliapeng2011) [reply src/plugins/intel_cpu/tests/unit/graph/inplace_resolve_io.cpp]: There should be a set of "desired" directions, v.s. only one "undesired".

Comment (maxnick) [reply src/plugins/intel_cpu/tests/unit/graph/inplace_resolve_io.cpp]: To my understanding the may be only two either UP or DOWN. So obviously only one of them is desired and one undesired, isn't it?

Comment (ceciliapeng2011) [reply src/plugins/intel_cpu/tests/unit/graph/inplace_resolve_io.cpp]: There could be others - NONE. Some cases both NONE and DOWN are desiring. Some cases both NONE and UP are desiring.

File: src/plugins/intel_cpu/src/node.cpp
Comment (maxnick) [inline src/plugins/intel_cpu/src/node.cpp:1]: Please delete debug code.

File: src/plugins/intel_cpu/src/node.cpp [hunk 7]
```diff
@@ -1828,14 +1828,75 @@ void Node::resolveInPlaceDirection() {
...
                     initDescriptor(config);
+                } else if (InplaceDirectionType::NONE == result) {
+                    // resolve cyclic inplace to downstream instead of upstream for the node
+                    // when there is only one output referencing to the edges of it,
+                    // thus benefits zero-copy of outputs.
+                    size_t numConflicts = 0;
+
+                    // the parent node does not use inPlace memory, but it is an Input.
+                    if (Type::Input == pParent->getType() || Type::MemoryInput == pParent->getType()) {
+                        numConflicts = 2;  // early stop
```
Comment (maxnick) [inline src/plugins/intel_cpu/src/node.cpp:1847]: Why not simply call the following?
```cpp
        auto config = getSelectedPrimitiveDescriptor()->getConfig();
        config.inConfs[inpPort].inPlace(-1);
        initDescriptor(config);
        continue;
```

File: src/plugins/intel_cpu/tests/functional/custom/subgraph_tests/src/inplace_resolve_io.cpp [hunk 8]
```diff
@@ -0,0 +1,321 @@
...
+        // static
+        {{2, 10, 8}, {{2, 10, 8}}}, // X [batch_size, seq_length, input_size]
+        {{2, 1, 3}, {{2, 1, 3}}},   // H_t [batch_size, num_directions, hidden_size]
+        {{2}, {{2}}}                // [batch_size]
+    },
+    {
+        // dynamic
+        {{-1, 10, 8}, {{2, 10, 8}}}, // param0
+        {{-1, 1, 3}, {{2, 1, 3}}}, // param1
+        {{-1}, {{2}}},  // param2
```
Comment (maxnick) [inline src/plugins/intel_cpu/tests/functional/custom/subgraph_tests/src/inplace_resolve_io.cpp:149]: Please add at least one more shape to the target shapes set.

File: src/plugins/intel_cpu/tests/functional/custom/subgraph_tests/src/inplace_resolve_io.cpp [hunk 9]
```diff
@@ -0,0 +1,321 @@
...
+    {
+        // static
+        {{2, 10, 8}, {{2, 10, 8}}}, // X [batch_size, seq_length, input_size]
+        {{2, 1, 3}, {{2, 1, 3}}},   // H_t [batch_size, num_directions, hidden_size]
+        {{2}, {{2}}}                // [batch_size]
+    },
+    {
+        // dynamic
+        {{-1, 10, 8}, {{2, 10, 8}}}, // param0
+        {{-1, 1, 3}, {{2, 1, 3}}}, // param1
```
Comment (maxnick) [inline src/plugins/intel_cpu/tests/functional/custom/subgraph_tests/src/inplace_resolve_io.cpp:148]:
```suggestion
        {{1, -1, 3}, {{1, 2, 3}}}, // param1
```

File: src/plugins/intel_cpu/tests/functional/custom/subgraph_tests/src/inplace_resolve_io.cpp [hunk 10]
```diff
@@ -0,0 +1,321 @@
...
+namespace {
+
+const std::vector<std::vector<InputShape>> inputShapes = {
+    {
+        // static
+        {{2, 64}, {{2, 64}}}
+    },
+    {
+        // dynamic
+        {{2, -1}, {{2, 64}}}
```
Comment (maxnick) [inline src/plugins/intel_cpu/tests/functional/custom/subgraph_tests/src/inplace_resolve_io.cpp:303]: Could you please add at least one more target shape to the dynamic case?

**Review by maxnick — CHANGES_REQUESTED** [review]

File: src/plugins/intel_cpu/src/node.cpp (hunk 7 above)
Comment (ceciliapeng2011) [reply src/plugins/intel_cpu/src/node.cpp:1847]: sounds good. done

File: src/plugins/intel_cpu/src/node.cpp
Comment (ceciliapeng2011) [reply src/plugins/intel_cpu/src/node.cpp:1]: done

File: src/plugins/intel_cpu/tests/functional/custom/subgraph_tests/src/inplace_resolve_io.cpp (hunk 9 above)
Comment (ceciliapeng2011) [reply src/plugins/intel_cpu/tests/functional/custom/subgraph_tests/src/inplace_resolve_io.cpp:148]: the second dim of param1 is num_directions, which should be 1 and cannot be dynamic in this test model, as it is used to generate constant W, B, R nodes. I add another test shape which makes sequence_len to be dynamic. Hope it is satisfying, though I don't really think that is need. This PR impacts graph, and if it covers both static and dynamic case, that's enough.

File: src/plugins/intel_cpu/tests/functional/custom/subgraph_tests/src/inplace_resolve_io.cpp (hunk 10 above)
Comment (ceciliapeng2011) [reply src/plugins/intel_cpu/tests/functional/custom/subgraph_tests/src/inplace_resolve_io.cpp]: done

File: src/plugins/intel_cpu/tests/functional/custom/subgraph_tests/src/inplace_resolve_io.cpp (hunk 8 above)
Comment (ceciliapeng2011) [reply src/plugins/intel_cpu/tests/functional/custom/subgraph_tests/src/inplace_resolve_io.cpp]: done

**Review by maxnick — APPROVED** [review]
LGTM

**Comment by maxnick** [comment]
Waiting internal validation.

**Pull request merged**